File: Show and edit date only and time only in a DataGridView/[C#][VB]-Show and edit date only and time only in a DataGridView/C# and VB/WindowsFormsApplication1/Controls/TimeColumn.cs
using System;
using System.Windows.Forms;

public class TimeColumn : DataGridViewColumn
{
    public TimeColumn() : base()
    {
        base.CellTemplate = new CalendarCell1();
    }

    public override DataGridViewCell CellTemplate
    {
        get
        {
            return base.CellTemplate;
        }
        set
        {
            if (!((value == null)) && !(value.GetType().IsAssignableFrom(typeof(CalendarCell1))))
            {
                throw new InvalidCastException("Must be a CalendarCell");
            }
            base.CellTemplate = value;
        }
    }
}
public class CalendarCell1 : DataGridViewTextBoxCell
{
    public CalendarCell1()
    {
        this.Style.Format = "hh:mm tt";
    }
    public override void InitializeEditingControl(int rowIndex, object initialFormattedValue, DataGridViewCellStyle dataGridViewCellStyle)
    {
        base.InitializeEditingControl(rowIndex, initialFormattedValue, dataGridViewCellStyle);

        CalendarEditingControl1 ctl = (CalendarEditingControl1)DataGridView.EditingControl;
        if (this.RowIndex >= 0)
        {
            if (!(object.ReferenceEquals(this.Value, DBNull.Value)))
            {
                if (this.Value != null)
                {
                    if (this.Value != "")
                    {
                        try
                        {
                            ctl.Value = DateTime.Parse(this.Value.ToString());
                        }
                        catch (Exception ex)
                        {
                        }
                    }
                }
            }
        }
    }
    public override System.Type EditType
    {
        get
        {
            return typeof(CalendarEditingControl1);
        }
    }
    public override System.Type ValueType
    {
        get
        {
            return typeof(DateTime);
        }
    }
    public override object DefaultNewRowValue
    {
        get
        {
            return DateTime.Now.ToShortTimeString();
        }
    }
}
internal class CalendarEditingControl1 : DateTimePicker, IDataGridViewEditingControl
{
    private DataGridView dataGridViewControl;
    private bool valueIsChanged = false;
    private int rowIndexNum;

    public CalendarEditingControl1()
    {
        this.Format = DateTimePickerFormat.Time;
    }
    public object EditingControlFormattedValue
    {
        get
        {
            return this.Value.ToShortTimeString();
        }
        set
        {
            if (value is string)
            {
                this.Value = DateTime.Parse(System.Convert.ToString(value));
            }
        }
    }
    public object GetEditingControlFormattedValue(DataGridViewDataErrorContexts context)
    {
        return this.Value.ToShortTimeString();
    }
    public void ApplyCellStyleToEditingControl(DataGridViewCellStyle dataGridViewCellStyle)
    {
        this.Font = dataGridViewCellStyle.Font;
        this.ShowUpDown = true;
        this.CalendarForeColor = dataGridViewCellStyle.ForeColor;
        this.CalendarMonthBackground = dataGridViewCellStyle.BackColor;
    }

    public int EditingControlRowIndex
    {
        get
        {
            return rowIndexNum;
        }
        set
        {
            rowIndexNum = value;
        }
    }

    public bool EditingControlWantsInputKey(Keys key, bool dataGridViewWantsInputKey)
    {
        if (Keys.KeyCode == Keys.Left || Keys.KeyCode == Keys.Up || Keys.KeyCode == Keys.Down || Keys.KeyCode == Keys.Right || Keys.KeyCode == Keys.Home || Keys.KeyCode == Keys.End || Keys.KeyCode == Keys.PageDown || Keys.KeyCode == Keys.PageUp)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    public void PrepareEditingControlForEdit(bool selectAll)
    {
    }

    public bool RepositionEditingControlOnValueChange
    {
        get
        {
            return false;
        }
    }

    public DataGridView EditingControlDataGridView
    {
        get
        {
            return dataGridViewControl;
        }
        set
        {
            dataGridViewControl = value;
        }
    }

    public bool EditingControlValueChanged
    {
        get
        {
            return valueIsChanged;
        }
        set
        {
            valueIsChanged = value;
        }
    }

    public Cursor EditingControlCursor
    {
        get
        {
            return base.Cursor;
        }
    }

    protected override void OnValueChanged(System.EventArgs eventargs)
    {
        valueIsChanged = true;
        this.EditingControlDataGridView.NotifyCurrentCellDirty(true);
        base.OnValueChanged(eventargs);
    }
    Cursor IDataGridViewEditingControl.EditingPanelCursor
    {
        get
        {
            return this.IDataGridViewEditingControl_EditingPanelCursor;
        }
    }
    private Cursor IDataGridViewEditingControl_EditingPanelCursor
    {
        get //throw new Exception("The method or operation is not implemented."); }
        {
            return base.Cursor;
        }
    }
}

File: Show and edit date only and time only in a DataGridView/[C#][VB]-Show and edit date only and time only in a DataGridView/C# and VB/WindowsApplication1/Controls/TimeColumn.vb
Imports System
Imports System.Windows.Forms

Namespace UserControls
    Public Class TimeColumn
        Inherits DataGridViewColumn

        Public Sub New()
            MyBase.New()
            MyBase.CellTemplate = New CalendarCell1()
        End Sub

        Public Overrides Property CellTemplate() As DataGridViewCell
            Get
                Return MyBase.CellTemplate
            End Get
            Set(ByVal value As DataGridViewCell)
                If Not ((value Is Nothing)) AndAlso Not (value.GetType().IsAssignableFrom(GetType(CalendarCell1))) Then
                    Throw New InvalidCastException("Must be a CalendarCell")
                End If
                MyBase.CellTemplate = value
            End Set
        End Property
    End Class
    Public Class CalendarCell1
        Inherits DataGridViewTextBoxCell

        Public Sub New()
            Me.Style.Format = "hh:mm tt"
        End Sub
        Public Overrides Sub InitializeEditingControl(ByVal rowIndex As Integer, ByVal initialFormattedValue As Object, ByVal dataGridViewCellStyle As DataGridViewCellStyle)
            MyBase.InitializeEditingControl(rowIndex, initialFormattedValue, dataGridViewCellStyle)

            Dim ctl As CalendarEditingControl1 = CType(DataGridView.EditingControl, CalendarEditingControl1)
            If Me.RowIndex >= 0 Then
                If ((Not Object.ReferenceEquals(Me.Value, DBNull.Value))) Then
                    If Me.Value IsNot Nothing Then
                        If Me.Value IsNot String.Empty Then
                            Try
                                ctl.Value = Date.Parse(Me.Value.ToString())
                            Catch ex As Exception
                            End Try
                        End If
                    End If
                End If
            End If
        End Sub
        Public Overrides ReadOnly Property EditType() As System.Type
            Get
                Return GetType(CalendarEditingControl1)
            End Get
        End Property
        Public Overrides ReadOnly Property ValueType() As System.Type
            Get
                Return GetType(Date)
            End Get
        End Property
        Public Overrides ReadOnly Property DefaultNewRowValue() As Object
            Get
                Return Date.Now.ToShortTimeString()
            End Get
        End Property
    End Class
    Friend Class CalendarEditingControl1
        Inherits DateTimePicker
        Implements IDataGridViewEditingControl

        Private dataGridViewControl As DataGridView
        Private valueIsChanged As Boolean = False
        Private rowIndexNum As Integer

        Public Sub New()
            Me.Format = DateTimePickerFormat.Time
        End Sub
        Public Property EditingControlFormattedValue() As Object Implements IDataGridViewEditingControl.EditingControlFormattedValue
            Get
                Return Me.Value.ToShortTimeString()
            End Get
            Set(ByVal value As Object)
                If TypeOf value Is String Then
                    Me.Value = Date.Parse(System.Convert.ToString(value))
                End If
            End Set
        End Property
        Public Function GetEditingControlFormattedValue(ByVal context As DataGridViewDataErrorContexts) As Object Implements IDataGridViewEditingControl.GetEditingControlFormattedValue
            Return Me.Value.ToShortTimeString()
        End Function
        Public Sub ApplyCellStyleToEditingControl(ByVal dataGridViewCellStyle As DataGridViewCellStyle) Implements IDataGridViewEditingControl.ApplyCellStyleToEditingControl
            Me.Font = dataGridViewCellStyle.Font
            Me.ShowUpDown = True
            Me.CalendarForeColor = dataGridViewCellStyle.ForeColor
            Me.CalendarMonthBackground = dataGridViewCellStyle.BackColor
        End Sub

        Public Property EditingControlRowIndex() As Integer Implements IDataGridViewEditingControl.EditingControlRowIndex
            Get
                Return rowIndexNum
            End Get
            Set(ByVal value As Integer)
                rowIndexNum = value
            End Set
        End Property

        Public Function EditingControlWantsInputKey(ByVal key As Keys, ByVal dataGridViewWantsInputKey As Boolean) As Boolean Implements IDataGridViewEditingControl.EditingControlWantsInputKey
            If Keys.KeyCode = Keys.Left OrElse Keys.KeyCode = Keys.Up OrElse Keys.KeyCode = Keys.Down OrElse Keys.KeyCode = Keys.Right OrElse Keys.KeyCode = Keys.Home OrElse Keys.KeyCode = Keys.End OrElse Keys.KeyCode = Keys.PageDown OrElse Keys.KeyCode = Keys.PageUp Then
                Return True
            Else
                Return False
            End If
        End Function

        Public Sub PrepareEditingControlForEdit(ByVal selectAll As Boolean) Implements IDataGridViewEditingControl.PrepareEditingControlForEdit
        End Sub

        Public ReadOnly Property RepositionEditingControlOnValueChange() As Boolean Implements IDataGridViewEditingControl.RepositionEditingControlOnValueChange
            Get
                Return False
            End Get
        End Property

        Public Property EditingControlDataGridView() As DataGridView Implements IDataGridViewEditingControl.EditingControlDataGridView
            Get
                Return dataGridViewControl
            End Get
            Set(ByVal value As DataGridView)
                dataGridViewControl = value
            End Set
        End Property

        Public Property EditingControlValueChanged() As Boolean Implements IDataGridViewEditingControl.EditingControlValueChanged
            Get
                Return valueIsChanged
            End Get
            Set(ByVal value As Boolean)
                valueIsChanged = value
            End Set
        End Property

        Public ReadOnly Property EditingControlCursor() As Cursor
            Get
                Return MyBase.Cursor
            End Get
        End Property

        Protected Overrides Sub OnValueChanged(ByVal eventargs As System.EventArgs)
            valueIsChanged = True
            Me.EditingControlDataGridView.NotifyCurrentCellDirty(True)
            MyBase.OnValueChanged(eventargs)
        End Sub
        Private ReadOnly Property IDataGridViewEditingControl_EditingPanelCursor() As Cursor Implements IDataGridViewEditingControl.EditingPanelCursor
            Get 'throw new Exception("The method or operation is not implemented."); }
                Return MyBase.Cursor
            End Get
        End Property
    End Class
End Namespace